pennylane/beta/tapes/qubit_param_shift.py

```diff
@@ -0,0 +1,228 @@
+        self.analytic_pd = self.parameter_shift
+
+        # check if the quantum tape contains any variance measurements
+        self.var_mask = [m.return_type is qml.operation.Variance for m in self.measurements]
+
+        if any(self.var_mask):
+            # The tape contains variances.
+            # Set parameter_shift_var as the analytic_pd method
+            self.analytic_pd = self.parameter_shift_var
+
+            # Make a copy of the original measurements; we will be mutating them
+            # during the parameter shift method.
+            self._original_measurements = self._measurements.copy()
+
+            # Finally, store the locations of any variance measurements in the
+            # measurement queue.
+            self.var_idx = np.where(self.var_mask)[0]
+
+    def _grad_method(self, idx, use_graph=True, default_method="A"):
+        op = self._par_info[idx]["op"]
+
+        if op.grad_method == "F":
+            return "F"
+
+        return super()._grad_method(idx, use_graph=use_graph, default_method=default_method)
+
+    def jacobian(self, device, params=None, **options):
+        # The parameter_shift_var method needs to evaluate the circuit
+        # at the unshifted parameter values; these are stored in the
+        # self._evA attribute. Here, we set the value of the attribute to None
+        # before each Jacobian call, so that the expectation value is calculated only once.
+        self._evA = None
+        return super().jacobian(device, params, **options)
+
+    def parameter_shift(self, idx, device, params, **options):
+        r"""Partial derivative using the parameter-shift rule of a tape consisting of measurement
+        statistics that can be represented as expectation values of observables.
+
+        This includes tapes that output probabilities, since the probability of measuring a
+        basis state :math:`|i\rangle` can be written in the form of an expectation value:
+
+        .. math::
+
+            \mathbb{P}_{|i\rangle} = |\langle i | U(\mathbf(p)) | 0 \rangle|^2
+                = \langle 0 | U(\mathbf(p))^\dagger | i \rangle\langle i | U(\mathbf(p)) | 0 \rangle
+                = \mathbb{E}\left( | i \rangle\langle i | \right)
+
+        Args:
+            idx (int): trainable parameter index to differentiate with respect to
+            device (.Device, .QubitDevice): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): the quantum tape operation parameters
+
+        Keyword Args:
+            shift (float): the parameter shift value
+
+        Returns:
+            array[float]: 1-dimensional array of length determined by the tape output
+            measurement statistics
+        """
+        t_idx = list(self.trainable_params)[idx]
+        op = self._par_info[t_idx]["op"]
+        p_idx = self._par_info[t_idx]["p_idx"]
+
+        s = (
+            np.pi / 2

```

Comment:
I'm not sure I understand...are we setting the shift to pi/2 if there is no recipe in the operator? Wouldn't it be better to have `None` just mean that we cannot use parameter shift since it is not defined?

Reply:
I discovered during this PR that the `grad_recipe` property is pretty bad 😆 

I think two things can happen:

* `grad_recipe = None`: gate is not differentiable.

* `grad_recipe = [None, (c, s)]`: gate has two parameters, first one takes default shift/coefficients, second one has custom mutliplier/shift.

I've been thinking how we can improve this:

* We should change this behaviour to instead be of the form `grad_recipe = [[(c0, s0), (c1, s1), ...], [(c0, s0), ...]]`, each tuple corresponding to a different _term_ to allow for arbitrary linear combinations of terms per gate parameter.

* `grad_recipe` is _very_ unique to the parameter-shift rules. Should we make this explicit? What if there is a case in the future where we have a new differentiation method that has it's own grad-recipe? (the state differentiation formula comes to mind).

Reply:
Another thing I didn't really like was `grad_method=None/"A"/"F"`.

Reply:
never liked that myself either